// @iota-wiki/cli/dist/cli/src/docusaurus/config/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
module.exports = {
    title: 'IOTA Wiki',
    tagline: 'The complete reference for IOTA',
    url: 'https://wiki.iota.org',
    baseUrl: '/',
    favicon: 'img/favicon.ico',
    trailingSlash: true,
    organizationName: 'iota-wiki',
    projectName: 'iota-wiki',
    stylesheets: [
        'https://fonts.googleapis.com/css?family=Material+Icons',
        {
            href: 'https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css',
            type: 'text/css',
            integrity: 'sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM',
            crossorigin: 'anonymous',
        },
    ],
    plugins: ['plugin-image-zoom'],
    themes: [
        [
            '@docusaurus/theme-classic',
            {
                customCss: path_1.default.resolve(__dirname, '../../../../src/next/css/custom.css'),
            },
        ],
        path_1.default.resolve(__dirname, '../theme'),
    ],
    themeConfig: {
        image: 'img/shimmer-wiki.png',
        imageZoom: {
            selector: '.markdown :not(a) > img:not(.image-gallery-image):not(.image-gallery-thumbnail-image)',
            // Optional medium-zoom options
            // see: https://www.npmjs.com/package/medium-zoom#options
            options: {
                background: 'rgba(0, 0, 0, 0.6)',
            },
        },
        navbar: {
            hideOnScroll: true,
            logo: {
                alt: 'IOTA Wiki Logo',
                src: 'img/logo_next.svg',
                srcDark: 'img/logo_next_dark.svg',
            },
            items: [],
        },
    },
    staticDirectories: [path_1.default.resolve(__dirname, '../static')],
};
